### PR TITLE
Port ngram_repeat_block and ngram cleaning over units to SC

### PR DESCRIPTION
**Pull request by cndn**
To potentially address stuttering and avoid noise that blast users' ears.

These options are set to take no effect by default to ensure paper results reproducibility.

### CODE REVIEW

File: src/seamless_communication/models/inference/ngram_repeat_block_logits_processor.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+from fairseq2.generation.logits_processor import LogitsProcessor as LogitsProcessor
+from typing import List
+from torch import Tensor
+import torch
+
+
+class NGramRepeatBlockLogitsProcessor(LogitsProcessor):
```
Comment (kauterry) [inline src/seamless_communication/models/inference/ngram_repeat_block_logits_processor.py:7]: Can recommends renaming this to `NGramRepeatBlockProcessor`

> Neat! I was going to suggest leveraging Naji's LogitsProcessor, you already have it, nice job. One nit suggestion, maybe you can slightly shorten the name by dropping "Logits", i.e. NGramRepeatBlockProcessor.

File: src/seamless_communication/models/inference/ngram_repeat_block_logits_processor.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    def __init__(self, no_repeat_ngram_size: int) -> None:
+        self.no_repeat_ngram_size = no_repeat_ngram_size
+
+    def __call__(self, seqs: Tensor, lprobs: Tensor) -> None:
+        """Remove repeating n-gram tokens."""
+        batch_size, beam_size, vocab_size = lprobs.size()
+        step_nr = seqs.size(2) - 1
+        # (N, B, step_nr + 1) -> (N * B, step_nr + 1)
```
Comment (kauterry) [inline src/seamless_communication/models/inference/ngram_repeat_block_logits_processor.py:15]: from Can:
> nit: You can annotate this simply (N, B, S) to be consistent with the annotations in Seq2SeqGenerator.

File: src/seamless_communication/models/inference/ngram_repeat_block_logits_processor.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        :param lprobs: The next-step log probability reshaped to (N * B, V)
+        :param batch_size: The batch size.
+        :param beam_size: The beam size.
+        :param step_nr: Step number for decoding.
+
+        :returns:
+            modified lprobs tensor with banned tokens set to -inf
+        """
+        banned_tokens = [
+            torch.jit.annotate(List[int], []) for _ in range(batch_size * beam_size)
```
Comment (kauterry) [inline src/seamless_communication/models/inference/ngram_repeat_block_logits_processor.py:43]: from Can:
> As of today, we do not use any torchscript annotations in the codebase as they reduce the readability a lot. We will have a totally different approach in place of torchscript, so you might want to remove the annotate call here.

**Review by kauterry — APPROVED** [review]

File: src/seamless_communication/models/inference/translator.py [hunk 4]
```diff
@@ -88,25 +91,28 @@ def get_prediction(
         input_modality: Modality,
         output_modality: Modality,
         tgt_lang: str,
+        ngram_filtering: bool = False,
```
Comment (kauterry) [inline src/seamless_communication/models/inference/translator.py:94]: Do you want to surface this as an argparser arg in scripts/m4t/predict?

File: src/seamless_communication/models/inference/ngram_repeat_block_logits_processor.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        ]
```
Comment (kauterry) [inline src/seamless_communication/models/inference/ngram_repeat_block_logits_processor.py:44]:
```suggestion
        banned_tokens = [[] for _ in range(batch_size * beam_size)]

```

File: src/seamless_communication/models/inference/ngram_repeat_block_logits_processor.py (hunk 2 above)
Comment (kauterry) [inline src/seamless_communication/models/inference/ngram_repeat_block_logits_processor.py:15]:
```suggestion
        # (N, B, S) -> (N * B, S)
```

File: src/seamless_communication/models/inference/ngram_repeat_block_logits_processor.py (hunk 1 above)
Comment (kauterry) [inline src/seamless_communication/models/inference/ngram_repeat_block_logits_processor.py:7]:
```suggestion
class NGramRepeatBlockProcessor(LogitsProcessor):
```

File: src/seamless_communication/models/inference/translator.py [hunk 6]
```diff
@@ -18,6 +18,9 @@
 from fairseq2.typing import Device
 from torch import Tensor
 from enum import Enum, auto
+from seamless_communication.models.inference.ngram_repeat_block_logits_processor import (
+    NGramRepeatBlockLogitsProcessor,
```
Comment (kauterry) [inline src/seamless_communication/models/inference/translator.py:22]:
```suggestion
    NGramRepeatBlockProcessor,
```

File: src/seamless_communication/models/inference/translator.py [hunk 7]
```diff
@@ -88,25 +91,28 @@ def get_prediction(
...
             # need to adjust this since src_len is smaller for text.
             max_len_a = 25
         else:
             max_len_a = 1
-
+        text_opts = SequenceGeneratorOptions(beam_size=5, soft_max_seq_len=(1, 200))
+        unit_opts = SequenceGeneratorOptions(beam_size=5, soft_max_seq_len=(max_len_a, 50))
+        if ngram_filtering:
+            text_opts.logits_processor = NGramRepeatBlockLogitsProcessor(no_repeat_ngram_size=4)
+            unit_opts.logits_processor = NGramRepeatBlockLogitsProcessor(no_repeat_ngram_size=4)
```
Comment (kauterry) [inline src/seamless_communication/models/inference/translator.py:105]:
```suggestion
            text_opts.logits_processor = NGramRepeatBlockProcessor(no_repeat_ngram_size=4)
            unit_opts.logits_processor = NGramRepeatBlockProcessor(no_repeat_ngram_size=4)
```
